File: assignments/assignment3a/app/src/test/java/edu/vandy/simulator/managers/beings/completionService/Assignment_3A_ExecutorCompletionServiceMgrTest.java
package edu.vandy.simulator.managers.beings.completionService;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.InOrder;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import java.util.List;
import java.util.concurrent.CompletionService;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import edu.vandy.simulator.ReflectionHelper;
import edu.vandy.simulator.utils.Assignment;

import static junit.framework.Assert.fail;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.atLeastOnce;
import static org.mockito.Mockito.doCallRealMethod;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class Assignment_3A_ExecutorCompletionServiceMgrTest {
    private final int BEING_COUNT = 5;
    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();
    @Mock
    public CompletionService<BeingCallable> mCompletionServiceMock;
    @Mock
    Future<BeingCallable> mFutureMock;
    @Mock
    private ExecutorService mExecutorMock;
    @InjectMocks
    private ExecutorCompletionServiceMgr mManagerMock =
            mock(ExecutorCompletionServiceMgr.class);

    // Used to count lambda errors caught by try/catch block
    private int mErrorCount = 0;

    // Used to reset assignment types back to original value before each test is run
    private int assignmentTypes;

    @Before
    public void before() {
        assignmentTypes = Assignment.sTypes;
    }

    @After
    public void after() {
        Assignment.sTypes = assignmentTypes;
    }

    @Test(timeout = 4000)
    public void testNewBeing() {
        doCallRealMethod().when(mManagerMock).newBeing();

        // Call SUT.
        BeingCallable beingCallable = mManagerMock.newBeing();

        assertNotNull("newBeing should not return null.", beingCallable);
    }

    @Test(timeout = 4000)
    public void testRunSimulation() {
        doCallRealMethod().when(mManagerMock).runSimulation();

        // Call SUT.
        mManagerMock.runSimulation();

        InOrder inOrder = inOrder(mManagerMock);

        verify(mManagerMock).beginBeingThreadPool();
        verify(mManagerMock).awaitCompletionOfFutures();
        verify(mManagerMock).shutdownNow();

        inOrder.verify(mManagerMock).beginBeingThreadPool();
        inOrder.verify(mManagerMock).awaitCompletionOfFutures();
        inOrder.verify(mManagerMock).shutdownNow();
    }

    @Test(timeout = 4000)
    public void testClassFieldsExist() throws Exception {
        ExecutorCompletionServiceMgr instance = new ExecutorCompletionServiceMgr();
        ReflectionHelper.injectValueIntoFirstMatchingField(instance, null, CompletionService.class);
        ReflectionHelper.injectValueIntoFirstMatchingField(instance, null, ExecutorService.class);
    }

    @Test(timeout = 4000)
    public void testCreateExecutorService() {
        when(mManagerMock.createExecutorService()).thenCallRealMethod();
        ExecutorService executorService = mManagerMock.createExecutorService();
        assertNotNull(executorService);
    }

    @Test(timeout = 4000)
    public void testCreateExecutorCompletionService() {
        when(mManagerMock.createExecutorCompletionService(any())).thenCallRealMethod();
        CompletionService<BeingCallable> completionService =
                mManagerMock.createExecutorCompletionService(mExecutorMock);
        assertNotNull(completionService);
    }

    /**
     * Test common to both UNDERGRADUATE And GRADUATE Assignments.
     */
    @Test(timeout = 4000)
    public void testBeginBeingThreadPool() {
        List<BeingCallable> mockBeings = createMockBeingList(BEING_COUNT);
        when(mManagerMock.createExecutorService()).thenReturn(mExecutorMock);
        when(mManagerMock.createExecutorCompletionService(mExecutorMock)).thenReturn(mCompletionServiceMock);
        when(mManagerMock.getBeings()).thenReturn(mockBeings);
        when(mCompletionServiceMock.submit(any(BeingCallable.class))).thenReturn(mFutureMock);
        doCallRealMethod().when(mManagerMock).beginBeingThreadPool();

        // Call SUT.
        mManagerMock.beginBeingThreadPool();

        verify(mManagerMock, times(1)).createExecutorService();
        verify(mManagerMock, times(1)).createExecutorCompletionService(mExecutorMock);
        verify(mManagerMock, times(1)).getBeings();
        verify(mCompletionServiceMock, times(mockBeings.size())).submit(any(BeingCallable.class));

        ReflectionHelper.findFirstMatchingFieldValue(mManagerMock, CompletionService.class);
        ReflectionHelper.findFirstMatchingField(mManagerMock, ExecutorService.class);
    }

    /**
     * Test common to both UNDERGRADUATE And GRADUATE Assignments.
     */
    @Test(timeout = 4000)
    public void testAwaitCompletionOfFuturesUndergraduates() throws Exception {
        BeingCallable beingCallableMock = mock(BeingCallable.class);
        List<BeingCallable> mockBeings = createMockBeingList(BEING_COUNT);
        when(mManagerMock.getBeings()).thenReturn(mockBeings);
        when(mManagerMock.getBeingCount()).thenReturn(mockBeings.size());
        when(mCompletionServiceMock.take()).thenReturn(mFutureMock);
        when(mFutureMock.get()).thenReturn(beingCallableMock);

        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mExecutorMock, ExecutorService.class);
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mCompletionServiceMock, CompletionService.class);

        doCallRealMethod().when(mManagerMock).awaitCompletionOfFutures();

        // Call SUT
        mManagerMock.awaitCompletionOfFutures();

        verify(mManagerMock, atLeastOnce()).getBeingCount();
        verify(mCompletionServiceMock, times(BEING_COUNT)).take();
        verify(mFutureMock, times(BEING_COUNT)).get();
    }

    @Test(timeout = 4000)
    public void testAwaitCompletionOfFuturesWithExceptionUndergraduates() throws Exception {
        if (!Assignment.testType(Assignment.UNDERGRADUATE)) {
            System.out.println("Skipping testAwaitCompletionOfFutures (undergraduate only test)");
            return;
        }

        List<BeingCallable> mockBeings = createMockBeingList(BEING_COUNT);
        when(mManagerMock.getBeings()).thenReturn(mockBeings);
        when(mManagerMock.getBeingCount()).thenReturn(mockBeings.size());
        when(mCompletionServiceMock.take()).thenReturn(mFutureMock);
        when(mFutureMock.get()).thenThrow(new RuntimeException("Mock exception"));

        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mExecutorMock, ExecutorService.class);
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mCompletionServiceMock, CompletionService.class);

        doCallRealMethod().when(mManagerMock).awaitCompletionOfFutures();

        try {
            // Call SUT
            mManagerMock.awaitCompletionOfFutures();
        } catch (Throwable e) {
            fail("Method should not throw an exception");
        }

        verify(mManagerMock, atLeastOnce()).getBeingCount();
        verify(mCompletionServiceMock, times(1)).take();
        verify(mFutureMock, times(1)).get();
    }

    @Test(timeout = 4000)
    public void testAwaitCompletionOfFuturesWithExceptionGraduates() throws Exception {
        if (!Assignment.testType(Assignment.GRADUATE)) {
            System.out.println("Skipping testAwaitCompletionOfFutures (graduate only test)");
            return;
        }

        List<BeingCallable> mockBeings = createMockBeingList(BEING_COUNT);
        when(mManagerMock.getBeings()).thenReturn(mockBeings);
        when(mManagerMock.getBeingCount()).thenReturn(mockBeings.size());
        when(mCompletionServiceMock.take()).thenReturn(mFutureMock);
        when(mFutureMock.get()).thenThrow(new RuntimeException("Mock exception"));

        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mExecutorMock, ExecutorService.class);
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mCompletionServiceMock, CompletionService.class);

        doCallRealMethod().when(mManagerMock).awaitCompletionOfFutures();

        try {
            // Call SUT
            mManagerMock.awaitCompletionOfFutures();
            fail("Method should throw an exception");
        } catch (RuntimeException e) {
        }

        verify(mManagerMock, atLeastOnce()).getBeingCount();
        verify(mCompletionServiceMock, times(1)).take();
        verify(mFutureMock, times(1)).get();
    }

    @Test(timeout = 4000)
    public void testShutdownNow() throws IllegalAccessException {
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mExecutorMock, ExecutorService.class);

        doCallRealMethod().when(mManagerMock).shutdownNow();

        // Call SUT.
        mManagerMock.shutdownNow();

        verify(mExecutorMock, times(1)).shutdownNow();
    }

    private List<Future<BeingCallable>> createMockFutureList(int count, boolean mockGet) {
        return IntStream.rangeClosed(1, count)
                .mapToObj(unused -> {
                    @SuppressWarnings("unchecked")
                    Future<BeingCallable> futureMock =
                            (Future<BeingCallable>) mock(Future.class);
                    try {
                        if (mockGet) {
                            when(futureMock.get()).thenReturn(mock(BeingCallable.class));
                        }
                    } catch (Exception e) {
                    }
                    return futureMock;
                })
                .collect(Collectors.toList());
    }

    private List<BeingCallable> createMockBeingList(int count) {
        return IntStream.rangeClosed(1, count)
                .mapToObj(unused -> {
                    BeingCallable being = mock(BeingCallable.class);
                    when(being.call()).thenReturn(being);
                    return being;
                })
                .collect(Collectors.toList());
    }
}

File: assignments/assignment4/app/src/test/java/edu/vandy/simulator/managers/beings/completionService/Assignment_3A_ExecutorCompletionServiceMgrTest.kt
package edu.vandy.simulator.managers.beings.completionService

import admin.AssignmentTests
import admin.ReflectionHelper
import com.nhaarman.mockitokotlin2.*
import edu.vandy.simulator.utils.Assignment
import org.junit.Assert
import org.junit.Test
import org.mockito.ArgumentMatchers
import org.mockito.InjectMocks
import org.mockito.Mock
import java.util.concurrent.CompletionService
import java.util.concurrent.ExecutorService
import java.util.concurrent.Future

class Assignment_3A_ExecutorCompletionServiceMgrTest : AssignmentTests() {
    companion object {
        private const val BEING_COUNT = 5
    }

    @Mock
    private lateinit var mCompletionServiceMock: CompletionService<BeingCallable>

    @Mock
    private lateinit var mExecutorMock: ExecutorService

    @Mock
    private lateinit var mFutureMock: Future<BeingCallable>

    @InjectMocks
    private var mManagerMock: ExecutorCompletionServiceMgr = mock()

    class SimulatedException : RuntimeException("Simulated exception")

    @Test
    fun testNewBeing() {
        doCallRealMethod().whenever(mManagerMock).newBeing()

        // Call SUT.
        val beingCallable = mManagerMock.newBeing()

        Assert.assertNotNull("newBeing should not return null.", beingCallable)
    }

    @Test
    fun testRunSimulation() {
        doCallRealMethod().whenever(mManagerMock).runSimulation()

        // Call SUT.
        mManagerMock.runSimulation()

        val inOrder = inOrder(mManagerMock)
        inOrder.verify(mManagerMock).beginBeingThreadPool()
        inOrder.verify(mManagerMock).awaitCompletionOfFutures()
        inOrder.verify(mManagerMock).shutdownNow()
    }

    @Test
    fun testClassFieldsExist() {
        val instance = ExecutorCompletionServiceMgr()
        ReflectionHelper.injectValueIntoFirstMatchingField(instance, null, CompletionService::class.java)
        ReflectionHelper.injectValueIntoFirstMatchingField(instance, null, ExecutorService::class.java)
    }

    @Test
    fun testCreateExecutorService() {
        whenever(mManagerMock.createExecutorService()).thenCallRealMethod()

        // Call SUT.
        val executorService = mManagerMock.createExecutorService()

        Assert.assertNotNull(executorService)
    }

    @Test
    fun testCreateExecutorCompletionService() {
        whenever(mManagerMock.createExecutorCompletionService(ArgumentMatchers.any())).thenCallRealMethod()

        // Call SUT.
        val completionService = mManagerMock.createExecutorCompletionService(mExecutorMock)

        Assert.assertNotNull(completionService)
    }

    /**
     * Test common to both UNDERGRADUATE And GRADUATE Assignments.
     */
    @Test
    fun testBeginBeingThreadPool() {
        val mockBeings = createMockBeingList(BEING_COUNT)
        whenever(mManagerMock.createExecutorService()).thenReturn(mExecutorMock)
        whenever(mManagerMock.createExecutorCompletionService(mExecutorMock)).thenReturn(mCompletionServiceMock)
        whenever(mManagerMock.beings).thenReturn(mockBeings)
        whenever(mCompletionServiceMock.submit(ArgumentMatchers.any(BeingCallable::class.java))).thenReturn(mFutureMock)
        doCallRealMethod().whenever(mManagerMock).beginBeingThreadPool()

        // Call SUT.
        mManagerMock.beginBeingThreadPool()

        verify(mManagerMock, times(1)).createExecutorService()
        verify(mManagerMock, times(1)).createExecutorCompletionService(mExecutorMock)
        verify(mManagerMock, times(1)).beings
        verify(mCompletionServiceMock, times(mockBeings.size)).submit(ArgumentMatchers.any(BeingCallable::class.java))
        ReflectionHelper.findFirstMatchingFieldValue<Any>(mManagerMock, CompletionService::class.java)
        ReflectionHelper.findFirstMatchingField(mManagerMock, ExecutorService::class.java)
    }

    /**
     * Test common to both UNDERGRADUATE And GRADUATE Assignments.
     */
    @Test
    fun testAwaitCompletionOfFuturesUndergraduates() {
        val beingCallableMock = mock<BeingCallable>()
        val mockBeings = createMockBeingList(BEING_COUNT)
        val size = mockBeings.size
        whenever(mManagerMock.beings).thenReturn(mockBeings)
        whenever(mManagerMock.beingCount).thenReturn(size)
        whenever(mCompletionServiceMock.take()).thenReturn(mFutureMock)
        whenever(mFutureMock.get()).thenReturn(beingCallableMock)
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mExecutorMock, ExecutorService::class.java)
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mCompletionServiceMock, CompletionService::class.java)
        doCallRealMethod().whenever(mManagerMock).awaitCompletionOfFutures()

        // Call SUT
        mManagerMock.awaitCompletionOfFutures()

        try {
            verify(mManagerMock, atLeast(2)).beingCount
        } catch (t: Throwable) {
            verify(mockBeings, atLeastOnce()).size
        }

        verify(mCompletionServiceMock, times(BEING_COUNT)).take()
        verify(mFutureMock, times(BEING_COUNT)).get()
    }

    @Test
    fun testAwaitCompletionOfFuturesWithExceptionUndergraduates() {
        if (!Assignment.testType(Assignment.UNDERGRADUATE)) {
            println("Skipping testAwaitCompletionOfFutures (undergraduate only test)")
            return
        }
        val mockBeings = createMockBeingList(BEING_COUNT)
        val size = mockBeings.size
        whenever(mManagerMock.beings).thenReturn(mockBeings)
        whenever(mManagerMock.beingCount).thenReturn(size)
        whenever(mCompletionServiceMock.take()).thenReturn(mFutureMock)
        whenever(mFutureMock.get()).thenThrow(SimulatedException())
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mExecutorMock, ExecutorService::class.java)
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mCompletionServiceMock, CompletionService::class.java)
        doCallRealMethod().whenever(mManagerMock).awaitCompletionOfFutures()

        // Call SUT
        assertThrows<RuntimeException> { mManagerMock.awaitCompletionOfFutures() }

        try {
            verify(mManagerMock, atLeast(2)).beingCount
        } catch (t: Throwable) {
            verify(mockBeings, atLeastOnce()).size
        }

        verify(mCompletionServiceMock, times(1)).take()
        verify(mFutureMock, times(1)).get()
    }

    @Test
    fun testAwaitCompletionOfFuturesWithExceptionGraduates() {
        if (!Assignment.testType(Assignment.GRADUATE)) {
            println("Skipping testAwaitCompletionOfFutures (graduate only test)")
            return
        }
        val mockBeings = createMockBeingList(BEING_COUNT)
        val size = mockBeings.size
        whenever(mManagerMock.beings).thenReturn(mockBeings)
        whenever(mManagerMock.beingCount).thenReturn(size)
        whenever(mCompletionServiceMock.take()).thenReturn(mFutureMock)
        whenever(mFutureMock.get()).thenThrow(SimulatedException())
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mExecutorMock, ExecutorService::class.java)
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mCompletionServiceMock, CompletionService::class.java)
        doCallRealMethod().whenever(mManagerMock).awaitCompletionOfFutures()

        // Call SUT
        assertThrows<RuntimeException> { mManagerMock.awaitCompletionOfFutures() }

        try {
            verify(mManagerMock, atLeast(2)).beingCount
        } catch (t: Throwable) {
            verify(mockBeings, atLeastOnce()).size
        }

        verify(mCompletionServiceMock, times(1)).take()
        verify(mFutureMock, times(1)).get()
    }

    @Test
    fun testShutdownNow() {
        ReflectionHelper.injectValueIntoFirstMatchingField(
                mManagerMock, mExecutorMock, ExecutorService::class.java)
        doCallRealMethod().whenever(mManagerMock).shutdownNow()

        // Call SUT.
        mManagerMock.shutdownNow()

        verify(mExecutorMock, times(1)).shutdownNow()
    }

    private fun createMockBeingList(count: Int): List<BeingCallable> {
        val list = (1..count).map {
            mock<BeingCallable>()
        }

        return spy(list)
    }
}